src/linpred.jl

```diff
@@ -336,37 +422,85 @@ function modelframe(f::FormulaTerm, data, contrasts::AbstractDict, ::Type{M}) wh
     f, modelcols(f, data)
 end
 
-modelmatrix(obj::LinPredModel) = obj.pp.X
+modelframe(obj::LinPredModel) = obj.fr
+
+modelmatrix(obj::LinPredModel; weighted::Bool=isweighted(obj)) = modelmatrix(obj.pp; weighted=weighted)
+
+function modelmatrix(pp::LinPred; weighted::Bool=isweighted(pp))
+    Z = if weighted
+        Diagonal(sqrt.(pp.wts))*pp.X
+    else
+        pp.X
+    end
+    return Z
+end
+
+function leverage(x::LinPredModel) 
+    h = vec(leverage(x.pp))
+    hasfield(typeof(x.rr), :wrkwt) ? x.rr.wrkwt.*h : x.rr.wts.*h
+end
+
+function leverage(pp::DensePredChol{T,<:CholeskyPivoted}) where T
+    X = modelmatrix(pp; weighted=false)
+    rnk = rank(pp.chol)
+    A = GLM.inverse(pp)
+    p = pp.chol.p[1:rnk]
+    diag(X[:,p]*A[p,p]*X[:,p]')
+    # sum(x->x^2, view(X, :, p)/view(pp.chol.U, p, p), dims=2)
+end
+
+function leverage(pp::DensePredChol{T,<:Cholesky}) where T
+    X = modelmatrix(pp; weighted=false)
+    @show X/pp.chol.U
+    sum(x -> x^2, X/pp.chol.U, dims=2)
+end
+
+function leverage(pp::DensePredQR{T,<:QRPivoted}) where T
+    X = modelmatrix(pp; weighted=false)
+    ch = pp.qr
+    rnk = rank(ch.R)
+    p = invperm(ch.p)[1:rnk]
+    sum(x -> x^2, view(X, :, 1:rnk)/view(ch.R, p, p), dims=2)
+end
+
+function leverage(pp::DensePredQR{T,<:QRCompactWY}) where T
+    X = modelmatrix(pp; weighted=false)
+    sum(x -> x^2, X/pp.qr.R, dims=2)
+end
+
 response(obj::LinPredModel) = obj.rr.y
 
 fitted(m::LinPredModel) = m.rr.mu
 predict(mm::LinPredModel) = fitted(mm)
-residuals(obj::LinPredModel) = residuals(obj.rr)
 
 function StatsModels.formula(obj::LinPredModel)
     obj.formula === nothing && throw(ArgumentError("model was fitted without a formula"))
     return obj.formula
 end
 
+residuals(obj::LinPredModel; weighted::Bool=false) = residuals(obj.rr; weighted=weighted)
+
 """
     nobs(obj::LinearModel)
     nobs(obj::GLM)
 
 For linear and generalized linear models, returns the number of rows, or,
 when prior weights are specified, the sum of weights.
 """
-function nobs(obj::LinPredModel)
-    if isempty(obj.rr.wts)
-        oftype(sum(one(eltype(obj.rr.wts))), length(obj.rr.y))
-    else
-        sum(obj.rr.wts)
-    end
-end
+nobs(obj::LinPredModel) = nobs(obj.rr)
+
+weights(obj::RegressionModel) = weights(obj.model)
```

Comment:
This is type piracy and no longer needed anyway in git master as we don't use `TableRegressionModel` anymore.
```suggestion
```